feat(masthead-v2.1): faceted (fka "tabbed") megamenu design updates

jkaeser opened the pull request:
### Related Ticket(s)

### Description

Updates component code and styles to make "faceted" megamenus comply with the 2.1 design specifications.

See the design & functional specs: https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3205%3A192891

### Changelog

**New**

- {{new thing}}

**Changed**

- Left and right columns in desktop megamenus scroll independently of one another.
- Divider line between left and right column only fades at the bottom if there is a vertically scrollable column.

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "test: e2e": Codesandbox examples and e2e integration tests -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

RichKummer commented:
Hey @jkaeser - saw that this PR didn't build correctly for web components. Is this change still being worked on?

jkaeser commented:
> Hey @jkaeser - saw that this PR didn't build correctly for web components. Is this change still being worked on?

Hey, @RichKummer. Yep, I'm still working through this one. Shouldn't take too much longer...

jwitkowski79 commented:
@jkaeser Hello! I think there may be a bug with the menu. I can't get the menu to display a dropdown in the L0. When I click on a a menu item in the L0, nothing happens. See attached video. Let me know if this is an error on my end. 

https://user-images.githubusercontent.com/2040791/213199066-3e40c914-cfdb-46d6-8a5b-0c2e8abb7591.mp4

andy-blum commented:
Seeing a couple of things here that seem "off", but I'd like a designer to confirm:

## Long headings wrap text, and the arrow is pushed too far right.

The "Featured" heading does not wrap, and the arrow is properly spaced.

<img width="350" alt="Screenshot 2023-01-18 at 9 34 53 AM" src="https://user-images.githubusercontent.com/25532785/213200062-d2b3c2dd-a408-4ee1-bbe7-b18622674578.png">

"Business Automation" wraps lines, and the arrow is spaced out too wide.

<img width="358" alt="Screenshot 2023-01-18 at 9 34 59 AM" src="https://user-images.githubusercontent.com/25532785/213200068-f7c858a1-2de3-4f5a-9980-109f98e2bedd.png">

## (Out of scope? Storybook only?) There are two skipnavs
 
<img width="355" alt="Screenshot 2023-01-18 at 9 35 32 AM" src="https://user-images.githubusercontent.com/25532785/213202988-f23d96bc-6ca4-468c-9f5b-202afe1bfc14.png">
<img width="288" alt="Screenshot 2023-01-18 at 9 35 43 AM" src="https://user-images.githubusercontent.com/25532785/213202991-2c42ceed-a05b-4752-bef8-4962e0f8c6aa.png">

## Megapanels with only one link-group overflow and do not wrap

I would assume that long lists like this should wrap before causing overflow.

<img width="1404" alt="Screenshot 2023-01-18 at 9 37 31 AM" src="https://user-images.githubusercontent.com/25532785/213203184-094f63ba-ec3a-4893-ab87-d202db5c8e85.png">

Removing wrapper divs/changing to `display:contents` allows links to wrap. We can prevent them from breaking by setting the link components to `display:block` and `break-inside:avoid`

<img width="1060" alt="Screenshot 2023-01-18 at 9 58 04 AM" src="https://user-images.githubusercontent.com/25532785/213205271-1b53dc9c-af76-45cb-8634-988b3cbac66d.png">

andy-blum commented on packages/styles/scss/components/masthead/_masthead-megamenu.scss:
```diff
@@ -161,7 +191,7 @@
     }
 
-    .#{$prefix}--masthead__megamenu--hasHighlights {
+    .#{$prefix}--masthead__megamenu--hasSidebar {
```
nit: we probably shouldn't mix BEM/snake-case with pascalCase here.

andy-blum commented on packages/services-store/src/types/translateAPI.ts:
```diff
@@ -253,3 +251,28 @@ export interface TranslateAPIState {
+export interface MegapanelLinkGroup {
+  heading?: BasicLink;
+  links?: BasicLink[];
+}
```
Note to future-us: We'll need to consider how to deploy the updates to the masthead with the updated JSON files as close to simultaneously as possible.

andy-blum commented on packages/web-components/src/components/masthead/__stories__/masthead.stories.ts:
```diff
@@ -16,7 +16,7 @@ import textNullable from '../../../../.storybook/knob-text-nullable';
 import '../masthead-container';
 import styles from './masthead.stories.scss';
-import { mastheadLinks as links, mastheadL1Data, logoData } from './links';
+import { NEWmastheadLinks as links, mastheadL1Data, logoData } from './links';
```
nit: Should `NEW` be all caps like this?

andy-blum commented:
@jwitkowski79 We're changing the structure of the JSON that feeds this component. For now you'll have to make sure you enable the "use mock data" knob. [Direct Link](https://ibmdotcom-webcomponents.s3.us-east.cloud-object-storage.appdomain.cloud/deploy-previews/9829/iframe.html?args=&knob-platform%20name%20(platform)=&knob-show%20the%20profile%20functionality%20(has-profile)=true&knob-show%20the%20search%20functionality%20(has-search)=true&knob-search%20placeholder%20(searchPlaceholder)=Search%20all%20of%20IBM&knob-selected%20menu%20item%20(selected-menu-item)=Consulting%20&%20Services=&knob-The%20user%20authenticated%20status%20(user-status)=Unauthenticated&knob-use%20mock%20nav%20data%20(use-mock)=true&id=components-masthead--with-v-2-data&viewMode=story)

jwitkowski79 commented:
@jkaeser: @andy-blum already mentioned some of the below but wanted to just reiterate that they are issues. Also, is mobile in a separate ticket? I think you had mentioned this but just want to double check

### Long headings wrap text, and the arrow is pushed too far right.
The heading level 1 wraps too soon in specific areas. Because of this, the arrow seems too far to the right. You can see this in Business Automation, Compute & Servers, Operating Systems and Security & Identity
![8F9FCEBD-C27F-485E-BDBF-8AC17A1517CF](https://user-images.githubusercontent.com/2040791/213237142-4699dce7-ebaf-4812-9524-1cf3a8707e38.png)

### Megapanels with only one link-group overflow and do not wrap
This use case probably won't happen very often because of content guidance but link groups should wrap as Andy indicates with with above 
![A76C36D2-A26B-46B1-BDD7-088A2108F8BF](https://user-images.githubusercontent.com/2040791/213237198-f4cdbadc-c492-45b4-ad49-f057ad25a13e.png)
solution

### Example data
I was curious if it's possible to show a couple other versions of sample data so we can properly QA other variations of how this menu could layout? If it's possible, we'd also like to see the following:
- [Heading level 2 with no description text](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3179%3A173943&t=kOzN5APjeq9KbyGF-1)
- [Heading level 2 with no description and link list (with no descriptions)](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3188%3A179024&t=kOzN5APjeq9KbyGF-1)
- [No Heading level 1, heading level 2 with no description and link list](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3179%3A173939&t=kOzN5APjeq9KbyGF-1)

### Left bottom rule needs to extend full width
The bottom rule in the left panel should extend fill width. Look like there's padding on the left and right side of it. [See Figma for specs](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node-id=3179%3A173939&t=kOzN5APjeq9KbyGF-1)

![B5249950-1EAF-45FB-BB3D-329249560117](https://user-images.githubusercontent.com/2040791/213240885-4dab5850-b9d2-461a-b46a-756e3b420e51.png)

I'd also love to get @oliviaflory eyes on this too.

jkaeser commented:
1. > andy-blum already mentioned some of the below but wanted to just reiterate that they are issues. Also, is mobile in a separate ticket? I think you had mentioned this but just want to double check
   > ### Left bottom rule needs to extend full width
   > The bottom rule in the left panel should extend fill width. Look like there's padding on the left and right side of it. [See 
Figma for specs](https://www.figma.com/file/35JfN0Sg4ltIIjM7jBDKMk/Navigation?node- id=3179%3A173939&t=kOzN5APjeq9KbyGF-1)

@jwitkowski79 Yes, sorry I didn't mention that before. I've updated the description to reflect that both mobile updates and changes to the "view all" links will happen in other tickets.

2. > ### Long headings wrap text, and the arrow is pushed too far right.
   > The heading level 1 wraps too soon in specific areas. Because of this, the arrow seems too far to the right. You can see this in Business Automation, Compute & Servers, Operating Systems and Security & Identity

Do you know what the maximum width for heading level 1 should be? I had made it one column to match the max width for its optional description.

3. > ### Megapanels with only one link-group overflow and do not wrap
   > This use case probably won't happen very often because of content guidance but link groups should wrap as Andy indicates with with above solution

I'll see what I can do here.

4. > ### Example data
   > I was curious if it's possible to show a couple other versions of sample data so we can properly QA other variations of how this menu could layout?

I'll add better demo content!

jwitkowski79 commented:
@jkaeser Hello! After discussing with @oliviaflory and Amanda, we agreed that the Heading level 1 should extend 8 columns total on desktop screen sizes to help avoid wrapping with the titles. We'll update the design specs to match this change. Let me know if you have any questions!

jwitkowski79 commented:
@jkaeser Hello! After have a discussion with @oliviaflory and Amanda about the faceted (tabbed) navigation, we came to the agreement that multiple H2's can appear in a single column. Even though this might be rare, it'll provide greater flexibility with the navigation itself. Even though an H2 does come with description text, we would recommend in the content guidance to never use it in this instance. I'm attaching a quick layout we did which will hopefully help give clarity. Please reach out if you have any questions!
<img width="894" alt="Screenshot 2023-01-19 at 11 25 56 AM (1)" src="https://user-images.githubusercontent.com/2040791/213527401-e79d3164-654f-46ae-bc67-bf14eb6a28a7.png">

andy-blum commented:
@jwitkowski79 the screenshot you attached shows the groups (heading & links) all being contained in the same column. Is this how it should act? Or is it possible for a group to break in the middle?

jkaeser commented:
@jwitkowski79 Alrighty, I've made some changes:

1. Heading 1 can now take up the full width of the megapanel before it wraps. The optional description still can only occupy a single column.
2. Single link groups in megapanels can now wrap into columns instead of creating one long column. This behavior only applies to single groups, and groups will only occupy a single column as soon as there are multiple of them in a megapanel.
3. I've redone the demo content to show the myriad display variations the megapanels can support. Let me know if there's another configuration you'd like to see represented.

> @jwitkowski79 the screenshot you attached shows the groups (heading & links) all being contained in the same column. Is this how it should act? Or is it possible for a group to break in the middle?

cc @andy-blum: Of note, the current implementation does not allow groups to break in the middle.

I believe I've addressed all the feedback at this time. Can you both take another look when you have a chance? Thanks!

jwitkowski79 commented:
@jkaeser This looks good. I would also love to get @oliviaflory eyes on this also if you can tag her as a reviewer in this masthead work

oliviaflory commented:
Hey @jkaeser @jwitkowski79 the desktop is looking great!

It doesn't look like the updated mobile styles are being applied, which is especially apparent when a heading 2 is present, is mobile in scope for this phase? Totally ok to address mobile in another PR, but I'd suggest not releasing on to live pages until the styles are added because it helps to hierarchy and is more clear what is clickable vs not.

**Seeing in PR**
<img width="547" alt="Screenshot 2023-01-23 at 3 04 49 PM" src="https://user-images.githubusercontent.com/45692622/214139187-ba8202ba-df87-4e6c-a4b0-6fe669229a22.png">

**Expected updated styling**
<img width="423" alt="Screenshot 2023-01-23 at 3 06 14 PM" src="https://user-images.githubusercontent.com/45692622/214139202-38c80455-20e8-45be-b588-0644f4f29f5b.png">

jkaeser commented:
> Hey @jkaeser @jwitkowski79 the desktop is looking great!
> 
> It doesn't look like the updated mobile styles are being applied, which is especially apparent when a heading 2 is present, is mobile in scope for this phase?

Thanks for taking a look, @oliviaflory! Mobile styles are being put together separately in another PR that Andy is working on.

Sounds like we've gotten design approval, so just waiting on @andy-blum's code feedback to the latest changes.
